### DISCUSSION TITLE
How can I get a dataflow graph (DAG) from a notebook via CLI or API?

### DISCUSSION THREAD

**Discussion by Meng6** [marimo-team/marimo#1877] [Q&A]
I checked the PyPI page and it mentioned that we could convert a Jupyter notebook to a marimo notebook with the CLI:
```
marimo convert your_notebook.ipynb > your_notebook.py
```

Similarly, is it possible to get a dataflow graph (DAG) from a notebook via CLI or API? For example, the input is a Jupyter notebook and the output is a JSON string:
```
{
    "cell1": {
        "usesVariable": [],
        "declaresVariable": ["v1", "v2"],
    },
    "cell2"{
        "usesVariable": ["v1"],
        "declaresVariable": ["v3"]
   },
    ...
}
```

Any suggestions? Thank you!

**Comment by mscolnick** [comment]
this is not currently possible, but seems like a useful artifact. we definitely have this knowledge. 

what is the use case? do you need the DAG by itself? if you are coming from a jupyter notebook, it is not guaranteed to be a DAG (could have cycles and re-declarations)

**Reply by Meng6** [reply]
Got it, thank you! I am thinking of generating dataflow graphs to understand existing codebases (Jupyter notebooks).

**Reply by mscolnick** [reply]
@Meng6 - is this something you'd be interested in contributing? shouldnt be much code after re-using some of the existing functions. 

we could add a CLI `marimo export dag --format=json` (not sure if we will have other formats). 

then you could `marimo convert your_notebook.ipynb > your_notebook.py && marimo export dag your_notebook.py --format=json`. or make it possible for `marimo export dag` to read from stdin

**Reply by Meng6** [reply]
Since the existing Jupyter Notebooks might not obey the assumptions such as `variables can’t be reassigned`. Can I reuse the marimo's code?

**Reply by mscolnick** [reply]
You could likely still get a dependency graph output, but may not be a DAG

**Reply by Meng6** [reply]
Cool! Does it mean marimo could extract variables per code cell without forcing the assumptions? If I contribute to this part, do I need to work on AST?

**Reply by mscolnick** [reply]
Im pretty sure (on mobile now, and need to check) we don't enforce those restrictions until actually running the code, which we would throw an error. Its likely no or very little change to get data graph without those restrictions

**Reply by Meng6** [reply]
That's great! Thank you! Could you please double check when you are free? If no need to work with AST, I'd like to contribute to it :) And if so, could you give me some guidance on how to do it?

**Reply by mscolnick** [reply]
@Meng6 I just doubled checked - you should not need to make any changes to the AST. 

You should be able to use `codegen.get_app` to get the `App` from a filename. you can refactor this to just pass contents if you want (probably nicer for a CLI if you can pass a file in from stdin). 

Once you have the `App`, you can traverse `App -> CellManager -> Cell[]` and each `Cell` has a set of `refs` and set of `defs`. Should be easy to create the serailized dag from there.

**Reply by Meng6** [reply]
Thank you! I tested the following case:

<img width="1308" alt="image" src="https://github.com/user-attachments/assets/497763df-3c5f-47ac-b7ff-c76726086baa">

For `lt`, I think it is used by cell-2 and cell-3. Is it a bug or is it supposed to be like this?

**Reply by mscolnick** [reply]
cell's are zero indexed. so it is used by cell1 and cell3, which is correct

**Reply by Meng6** [reply]
Isn't it used by `val = test(1, 3)`?

**Reply by mscolnick** [reply]
transitively, yes, but the DAG only shows "parent/children" not "ancestors/descendants". the later can be inferred

**Reply by Meng6** [reply]
I see, thank you!

**Reply by Meng6** [reply]
Thanks for all the help! I need to think about my project carefully & might contribute to Marimo in the future :)
Thanks for developing this fancy notebook!

**Comment by NukeMike** [comment]
I have created a notebook to access the dag graph within the notebook itself, and represent it trough a mo.mermaid component.
I had to handle the cells linear order as stored in mo._ast.cellmanaager
[Cell_manager.py](https://github.com/marimo-team/marimo/blob/main/marimo%2F_ast%2Fcell_manager.py#L48)
 
And [app.graph()](https://github.com/marimo-team/marimo/blob/main/marimo%2F_ast%2Fapp.py#L965)
If I recall well (on mobile now)

I found some useful classes and methods also in the _ai repo tools.

From within the notebook I could also load an external notebook and show it's graph.

My goal was multiple:
I wanted a graph view like the one in the sidebar, but more compact, just with cell number, cell custom name, and possibly function defined.
Have an embedded view in the notebook
Be able to export the graph view either as mermaid source or markdown with a mermaid code block that GitHub and vscode can already render (in vscode with an extension)

I purposedly created a graph to mermaid syntax function.
I also created a parent child / selector, to show the graph only of selected cells.

Then I started using more the minimap, (that does not fill the full width of the panel as the graph does)

Different views in the Marino native graph panel could be button enabled to show only titles or cell number.

My notebook could easily be adapted to run as a cli script.
